clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -494,10 +579,35 @@ private void handleDisconnections(List<ClientResponse> responses, long now) {
     /**
      * Record any newly completed connections
      */
-    private void handleConnections() {
+    private void handleConnections(long now) {
         for (String node : this.selector.connected()) {
             log.debug("Completed connection to node {}", node);
-            this.connectionStates.connected(node);
+            // Though the node is connected, we might not still be able to send requests. For instance,
+            // in case of SSL connection, SSL handshake happens after connection is established.
+            if (this.requiredApiVersions == null) {
+                this.connectionStates.ready(node);
+            } else {
+                this.connectionStates.checkingApiVersions(node);
+                nodesNeedingApiVersionsFetch.add(node);
+            }
+        }
+        maybeInitiateApiVersionsFetch(now);
+    }
+
+    private void maybeInitiateApiVersionsFetch(long now) {
+        if (this.requiredApiVersions == null)
+            return;
+
+        Iterator<String> iter = nodesNeedingApiVersionsFetch.iterator();
+        while (iter.hasNext()) {
+            String node = iter.next();
+            if (selector.isChannelReady(node) && inFlightRequests.canSendMore(node)) {
+                log.debug("Initiating API versions fetch from node {}.", node);
+                ClientRequest clientRequest = new ClientRequest(node, now, true,
+                        nextRequestHeader(ApiKeys.API_VERSIONS), ApiVersionsRequest.API_VERSIONS_REQUEST, null);
+                doSend(clientRequest, false, now);
```

Comment:
I initially thought to do that, but it is very likely that someone would have raised that as not required in review. However, now that is is raised, happy to add it. If there are no more reviews, I will shortly update the PR here.